text/0000-tuple-struct-self-ctor.md

```diff
@@ -0,0 +1,346 @@
+
+```rust
+struct Person(String, usize, usize);
+
+impl Person {
+    /// Make a newborn person.
+    fn newborn(name: String, ssn: usize) -> Self {
+        Self(name, ssn, 0)
+    }
+}
+```
+
+As with structs with named fields, you may also use `Self` when
+you are `impl`ing on a type alias of a struct as seen here:
+
+```rust
+struct FooBar(u8);
+
+type BarFoo = FooBar;
+
+impl Default for BarFoo {
+    fn default() -> Self {
+        Self(42) // <-- Not allowed before.
+    }
+}
+```
+
+## Patterns
+
+Currently, you can pattern match using `Self { .. }` on a named struct as in
+the following example:
+
+```rust
+struct Person {
+    ssn: usize,
+    age: usize
+}
+
+impl Person {
+    /// Make a newborn person.
+    fn newborn(ssn: usize) -> Self {
+        match { Self { ssn, age: 0 } } {
+            Self { ssn, age } // `Self { .. }` is permitted as a pattern!
+                => Self { ssn, age }
+        }
+    }
+}
+```
+
+This RFC extends this to tuple structs:
+
+```rust
+struct Person(usize, usize);
+
+impl Person {
+    /// Make a newborn person.
+    fn newborn(ssn: usize) -> Self {
+        match { Self(ssn, 0) } {
+            Self(ssn, age) // `Self(..)` is permitted as a pattern!
+                => Self(ssn, age)
+        }
+    }
+}
+```
+
+Of course, this redundant reconstruction is not recommended in actual code,
+but illustrates what you can do.
+
+## `Self` as a function pointer
+
+When you define a tuple struct today such as:
+
+```rust
+struct Foo<T>(T);
+
+impl<T> Foo<T> {
+    fn fooify_iter(iter: impl Iterator<Item = T>) -> impl Iterator<Item = Foo<T>> {
+        iter.map(Foo)
+    }
+}
+```
+
+you can use `Foo` as a function pointer typed at: `for<T> fn(T) -> T` as
+seen in the example above.
+
+This RFC extends that such that `Self` can also be used as a function pointer
+for tuple structs. Modifying the example above gives us:
+
+```rust
+impl<T> Foo<T> {
+    fn fooify_iter(iter: impl Iterator<Item = T>) -> impl Iterator<Item = Foo<T>> {
+        iter.map(Self)
+    }
+}
+```
+
+## Unit structs
+
+With this RFC, you can also use `Self` in pattern and expression contexts when
+dealing with unit structs. For example:
+
+```rust
+struct TheAnswer;
+
+impl Default for TheAnswer {
+    fn default() -> Self {
+        match { Self } { Self => Self }
+    }
+}
+```
+
+## Teaching the contents
+
+This RFC should not require additional effort other than spreading the
+news that this now is possible as well as the reference. The changes are
+seen as intuitive enough that it supports what the user already assumes
+should work and will probably try at some point.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Grammar
+
+Given:
+
+```
+%token SELF
+```
+
+which lexes `Self`, the following are legal productions in the language:
+
+```
+pat : ... // <-- The original grammar of `pat` prior to this RFC.
+    | SELF '(' ')'
+    | SELF '(' pat_tup ')'
```

Comment:
The grammar doesn't change, technically, it's still `PATH ( ... )` and `Self` is already a valid path.

Reply:
So the entire `## Grammar` section can be removed then?

Reply:
Yes.

Reply:
Aight; removed it :)